server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -315,10 +316,22 @@ public boolean offer(E e) {
      * A handler for rejected tasks that adds the specified element to this queue,
      * waiting if necessary for space to become available.
      */
-    static class ForceQueuePolicy implements XRejectedExecutionHandler {
+    static class ForceQueuePolicy extends EsRejectedExecutionHandler {
+
+        private final boolean rejectAfterShutdown;
+
+        /**
+         * @param rejectAfterShutdown indicates if {@link Runnable} should be rejected once the thread pool is shutting down
+         */
+        ForceQueuePolicy(boolean rejectAfterShutdown) {
+            this.rejectAfterShutdown = rejectAfterShutdown;
+        }
 
         @Override
         public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+            if (rejectAfterShutdown && executor.isShutdown()) {
+                throw newRejectedException(r, executor, true);
+            }
```

Comment:
After reading the code again I think you are right. I pushed a939ea6810864b466c28a9e2806e9bed81a9789b to add another test that executes tasks while concurrently shutting down the executor, making it more likely to have the race condition (on slow machines, it only fails 1 on 10K on my workstation though).

I reworked the logic in 0d89228af1656d4a336f294bb91c17fc91b31e7b to avoid the race condition, let me know what you think.